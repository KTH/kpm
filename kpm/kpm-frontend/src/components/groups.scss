@use "../vars" as *;

#kpm-6cf53 .kpm-collapsable-group {
  .kpm-link-list {
    margin: 0;

    li {
      margin: 0;
      padding: 0;
    }
  }

  summary {
    cursor: pointer;
  }
}

#kpm-6cf53 .kpm-dropdown-backdrop {
  position: fixed;
  cursor: pointer;
  z-index: 998;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

#kpm-6cf53 .kpm-dropdownmenu-group {
  position: relative;
  margin-bottom: 0; // Override KTH Style

  .kpm-link-list {
    // NOTE: Currently this popup menu is layered on top of the menubar
    // read comment for usePositionDropdown() for more info.
    position: fixed;
    z-index: 999;
    top: 0;
    left: 0;
    max-width: 95%;
    min-height: 200px; // Use a constant value in pixels, otherwise we have problems
    background-color: var(--kpmPaneBg);
    overflow-y: auto;
    box-shadow: 4px 4px 10px rgb(0 0 0 / 20%);
    visibility: hidden;
    opacity: 0;
    transition: opacity 0.1s ease-in;

    ul {
      padding: var(--kpmPad);

      @media #{$mblBreakpoint} {
        padding: 0;
      }
    }

    li.item {
      margin: 0;
      padding: 0 calc(2 * var(--kpmPad)) 0 0;

      @media #{$mblBreakpoint} {
        padding: calc(2 * var(--kpmPad)) calc(4 * var(--kpmPad));
      }
    }
  }

  summary {
    position: relative;
    cursor: context-menu;
    padding: 0; // Override KTH Style
    min-height: 0; // Override KTH Style
    padding-right: 1rem;
    list-style: none; // We use our own icon
  }

  // Hide arror in Safari and KTH Style
  & > summary::-webkit-details-marker,
  & > summary::before {
    display: none;
  }

  & > summary:first-of-type::after {
    content: "";
    background: $data-arrow-expand-black-svg no-repeat;
    width: 0.3rem;
    height: 0.5rem;
    position: absolute;
    top: 50%;
    right: 0;
    transform: translateY(-50%) rotate(0deg);
    transition: transform 0.1s ease-in-out;
  }

  &[open] > summary:first-of-type::after {
    transform: translateY(-50%) rotate(90deg);
  }
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-dropdownmenu-group {
    position: unset;
    flex-basis: 100%;
    font-size: 1rem;

    & > summary {
      font-size: 1em;
    }

    & > summary:first-of-type::after {
      width: 0.4rem;
      height: 0.65rem;
    }

    .kpm-link-list {
      position: fixed;
      max-width: unset;
      max-height: unset;
      width: 100% !important;
      height: 100vh !important;
      top: 0;
      left: 0;
      transform: translate(100%, 0) !important;
      transition: transform 0.25s ease-in-out;

      & > ul {
        font-size: 1.1em;
      }
    }
  }

  #kpm-6cf53 [open].kpm-dropdownmenu-group {
    .kpm-link-list.open {
      transform: translate(0, 0) !important;
    }
  }
}
