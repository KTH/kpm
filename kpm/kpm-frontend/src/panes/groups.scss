@import "../vars";

#kpm-6cf53 .kpm-groups {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: calc(4 * var(--kpmGap));
  $colCount: 3;
  $colWidth: calc(100% / $colCount - $colGap * ($colCount - 1) / $colCount);

  & > ul {
    display: flex;
    flex-flow: row wrap;
    gap: $rowGap $colGap;
    align-items: flex-start;
    width: 100%;

    & > .kpm-groups-item {
      display: flex;
      align-items: start;
      width: $colWidth;
      margin-bottom: 0;
      padding: var(--kpmPad);
      gap: $rowGap $colGap;

      .icon.star {
        color: var(--kpmIconStarBg);
        // Fix alignment error
        margin-bottom: 2px;

        &.active {
          color: var(--kpmIconStarActiveBg);
        }
      }
    }
  }
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-groups > ul > .kpm-groups-item {
    flex-basis: 100%;
    font-size: 1.1em;
  }
}
