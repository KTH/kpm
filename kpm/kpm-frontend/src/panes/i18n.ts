export type TLang = "sv" | "en";

export function i18n(
  strObj: Record<TLang, string> | string,
  lang: TLang = "sv"
) {
  if (typeof strObj === "object") {
    return strObj[lang];
  }

  // TODO: Create language dict for static string lookup
  return strObj;
}
