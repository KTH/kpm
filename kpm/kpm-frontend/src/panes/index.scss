@use "./teaching.scss";

.kpm-studies-course {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  gap: 0.5em;

  h2,
  h3 {
    font-size: 1em;
    margin: 0;
    padding: 0;
  }
}
