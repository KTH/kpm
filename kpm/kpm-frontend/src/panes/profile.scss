@use "../vars" as *;
@use "@kth/style/scss/tokens/typography";

#kpm-6cf53 div.kpm-profile {
  h3 {
    @include typography.font-heading-s;
    margin-block-start: 1rem;
    margin-block-end: 0.25rem;
  }

  h4 {
    @include typography.font-heading-xs;
    margin-block-start: 0.25rem;
    margin-block-end: 0.125rem;
  }
}
#kpm-6cf53 .kpm-profile .kpm-profile-link {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: $rowGap;

  display: flex;
  flex-flow: row nowrap;
  gap: $rowGap $colGap;
  align-items: center;

  img {
    height: 3em;
    width: 3em;
  }
}

#kpm-6cf53 .kpm-button.change-lang::after {
  display: inline-block;
  content: " ";
  width: 1em;
  height: 1em;
  background: $data-language-black-svg;
  margin: 0.2em 0 0.2em 1ch;
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-profile {
    & > .kpm-col {
      font-size: 1.1em;
    }
  }
}
