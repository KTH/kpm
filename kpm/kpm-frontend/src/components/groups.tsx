import React, {
  RefObject,
  useState,
  useEffect,
  useRef,
  MouseEvent,
} from "react";

import "./groups.scss";

export function CollapsableGroup({
  title,
  children,
  defaultOpen = false,
}: any) {
  return (
    <details className="kpm-collapsable-group">
      <summary>{title}</summary>
      <ul className="kpm-link-list">{children}</ul>
    </details>
  );
}

type TStyle = Record<string, string | undefined> | undefined;

type TDropdownMenuGroupProps = {
  title: string;
  className?: string;
  children: any;
  revealUp?: boolean;
  alignRight?: boolean;
  defaultOpen?: boolean;
  modal?: boolean; // Will not allow other interactions until closed
};

export function DropdownMenuGroup({
  title,
  className,
  children,
  revealUp = false,
  alignRight = false,
  defaultOpen = false,
  modal = false,
}: TDropdownMenuGroupProps) {
  const [open, setOpen] = useState(defaultOpen);
  const [dropdownStyle, setDropdownStyle]: [TStyle, Function] =
    useState(undefined);

  const detailsRef = useRef(null);
  const summaryRef = useRef(null);
  const dropdownRef = useRef(null);
  usePositionDropdown(
    detailsRef,
    summaryRef,
    dropdownRef,
    revealUp,
    alignRight,
    (newStyle: TStyle) => {
      if (newStyle === undefined || !isEqual(dropdownStyle, newStyle))
        setDropdownStyle(newStyle);
    }
  );

  // TODO: Listen to onBlur and onFocus on summary element

  let cls = "kpm-dropdownmenu-group";
  if (className !== undefined) {
    cls += ` ${className}`;
  }

  const doToggleOnMouseDown = (e: MouseEvent) => {
    if (summaryRef.current === e.target) {
      const currentOpen = open;
      setTimeout(() => {
        setOpen(!currentOpen);
      }, 100);
    }
  };

  const doOpenOnFocus = (event: any) => {
    const e = event as FocusEvent;
    if (e.target === summaryRef.current) {
      // Open on focus
      if (!open) {
        e.preventDefault();
        setOpen(true);
      }
    }
  };

  const doCloseOnBlur = (event: any) => {
    // Got type error when assigning (e: FocusEvent) => void to onBlur so using any for arg
    const e = event as FocusEvent;
    if (
      open &&
      !elementDescendentOf(e.relatedTarget as Node, detailsRef.current!)
    ) {
      e.preventDefault();
      setOpen(false);
    }
  };

  const doCloseOnBackropClick = () => {
    setOpen(false);
  };

  const _inner = (
    <div style={dropdownStyle} className="kpm-link-list">
      <ul ref={dropdownRef}>{children}</ul>
    </div>
  );

  const doSilenceDefaultDetailsOpen = (event: any) => {
    // Silence event so browser doesn't cause dropdown to toggle
    // state when onMouseDown toggle is performed. To avoid
    // toggle open + toggle close = no change
    const e = event as MouseEvent;
    if (e.target === summaryRef.current) {
      e.preventDefault();
    }
  };

  return (
    <details
      ref={detailsRef}
      className={cls}
      open={open}
      onClick={doSilenceDefaultDetailsOpen}
      onBlur={doCloseOnBlur}
    >
      <summary
        ref={summaryRef}
        onFocus={doOpenOnFocus}
        onMouseDown={doToggleOnMouseDown}
      >
        {title}
      </summary>
      {modal ? (
        <ModalBackdrop onClose={doCloseOnBackropClick}>{_inner}</ModalBackdrop>
      ) : (
        _inner
      )}
    </details>
  );
}

function elementDescendentOf(el: Node | null, parent: Node) {
  let tmpEl = el?.parentNode;
  while (tmpEl && (tmpEl as Element).tagName !== "body") {
    if (tmpEl === parent) return true;
    tmpEl = tmpEl.parentNode || null;
  }
  return false;
}

type TModalBackdropProps = {
  children?: JSX.Element;
  onClose?: () => void;
};
function ModalBackdrop({ children, onClose }: TModalBackdropProps) {
  const [oldOverflow, setOldOverflow] = useState("");

  useEffect(() => {
    setOldOverflow(document.body.style.overflow);
    document.body.style.overflow = "hidden";
    return () => {
      document.body.style.overflow = oldOverflow;
    };
  }, []);

  return (
    <div
      className="kpm-dropdown-backdrop"
      tabIndex={-1}
      onClick={() => {
        onClose && onClose();
      }}
    >
      {children}
    </div>
  );
}

function isEqual(a: TStyle, b: TStyle): boolean {
  if (a === undefined || b === undefined) return a === b;
  const keys = { ...a, ...b };

  for (const key of Object.keys(keys)) {
    if (a[key] !== b[key]) {
      return false;
    }
  }

  return true;
}

export function GroupItem({ className, children }: any) {
  let cls = "item";
  if (className !== undefined) {
    cls += ` ${className}`;
  }
  return <li className={cls}>{children}</li>;
}

const PADDING = 10;

/**
  Adjust size of dropdown to avoid overflow and switch side (vertical only) to allow
  best use of available screen realestate.

  We do this continuosly on animation frame becuase there are lots of layout changes
  that can affect the placement and don't generate events.

  Currently we aren't aware of the menubar so the popup needs to be layered on top to make
  sure we don't hide it with the menubar. This looks wierd if we aren't in modal mode
  (default) but not worth pursuing a fix for now.
 */
function usePositionDropdown(
  detailsRef: RefObject<HTMLElement | null>,
  summaryRef: RefObject<HTMLElement | null>,
  dropdownRef: RefObject<HTMLElement | null>,
  toTop: boolean = false,
  toRight: boolean = false,
  callback: Function
) {
  const requestRef = useRef(0);

  const calculate = () => {
    // Stop if unmounted
    const isOpen =
      (detailsRef.current as unknown as HTMLDetailsElement)?.open || false;
    const outer = summaryRef.current;
    const dropdown = dropdownRef.current;
    if (!isOpen || dropdown === null || outer === null) {
      callback(undefined);
      return (requestRef.current = requestAnimationFrame(calculate));
    }

    // 1. get viewport limits
    const viewport = window.visualViewport;
    const pageWidth = Math.round(viewport!.width);
    const pageHeight = Math.round(viewport!.height);
    // 2. get position and size of dropdown
    // https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect
    const rect = outer.getBoundingClientRect();
    const elTop = Math.round(rect.top);
    const elBottom = Math.round(rect.bottom);
    const elLeft = Math.round(rect.left);
    const elRight = Math.round(rect.right);
    const elHeight = Math.round(rect.height);
    const elWidth = Math.round(rect.width);

    const dropdownWidth = dropdown.scrollWidth;
    const dropdownHeight = dropdown.scrollHeight;

    // 3. Adjust placement and size depending on available space
    const spaceTop = elTop;
    const spaceBottom = pageHeight - elBottom;
    const spaceLeft = elRight;
    const spaceRight = pageWidth - elLeft;

    const cs = window.getComputedStyle(dropdown);
    const minHeight = getLargestCssValue(cs, ["min-height", "height"]) || 200;
    const minWidth = getLargestCssValue(cs, ["min-width", "width"]) || elWidth;

    // Should the dropdown open up or down? Depends on setting but also available space
    let placeTop;
    if (toTop) {
      if (spaceTop < minHeight + PADDING && spaceBottom > minHeight + PADDING) {
        // Flip to bottom because more space
        placeTop = false;
      } else {
        placeTop = true;
      }
    } else {
      if (spaceBottom < minHeight + PADDING && spaceTop > minHeight + PADDING) {
        // Flip to top beacuase more space
        placeTop = true;
      } else {
        placeTop = false;
      }
    }

    let deltaY; // may depend on actual height of dropdown
    let height = dropdownHeight;

    // Adjust size of dropdown if no space
    if (placeTop) {
      // Adjust space and placing for top position
      if (spaceTop < minHeight + PADDING) {
        height = minHeight;
      } else if (spaceTop < dropdownHeight + PADDING) {
        height = spaceTop - PADDING;
      }
      deltaY = elTop - height;
    } else {
      // Adjust space and placing for bottom position
      if (spaceBottom < minHeight + PADDING) {
        height = minHeight;
      } else if (spaceBottom < dropdownHeight + PADDING) {
        height = spaceBottom - PADDING;
      }
      deltaY = elBottom;
    }

    // Nudge placement on X axis depending on available space
    let placeRight;
    if (toRight) {
      if (spaceLeft < minWidth + PADDING && spaceRight > minWidth + PADDING) {
        // Flip to left aligned because more space
        placeRight = false;
      } else {
        placeRight = true;
      }
    } else {
      if (spaceRight < minWidth + PADDING && spaceLeft > minWidth + PADDING) {
        // Flip to right aligned because more space
        placeRight = true;
      } else {
        placeRight = false;
      }
    }

    let deltaX = placeRight ? elRight - dropdownWidth : elLeft;
    let width = minWidth;
    if (toRight) {
      if (deltaX < 0) {
        deltaX = elRight >= 0 ? 0 : elRight; // perhaps we shouldn't let nudge so far
      } else if (deltaX + dropdownWidth > pageWidth) {
        deltaX = pageWidth - dropdownWidth;
        if (deltaX + dropdownWidth < elLeft) {
          deltaX = elLeft - dropdownWidth; // perhaps we shouldn't let nudge so far
        }
      }
    } else {
      if (deltaX < 0) {
        deltaX = -elLeft;
        if (deltaX > elRight) {
          deltaX = elRight; // perhaps we shouldn't let nudge so far
        }
      } else if (deltaX + dropdownWidth > pageWidth) {
        deltaX = pageWidth - dropdownWidth;
        if (deltaX + dropdownWidth < elLeft) {
          deltaX = elLeft - dropdownWidth; // perhaps we shouldn't let nudge so far
        }
      }
    }

    let newTransform: any = {
      transform: `translate(${deltaX}px, ${deltaY}px)`,
      visibility: "visible",
      opacity: "1",
    };
    if (width !== undefined) {
      newTransform["width"] = `${width}px`;
    }
    if (height !== undefined) {
      newTransform["height"] = `${height}px`;
    }
    callback(newTransform);
    requestRef.current = requestAnimationFrame(calculate);
  };

  // Continue polling until unmounted
  useEffect(() => {
    requestRef.current = requestAnimationFrame(calculate);
    return () => cancelAnimationFrame(requestRef.current);
  }, []);
}

// Utils
function getLargestCssValue(cs: CSSStyleDeclaration, props: string[]) {
  let outp = 0;
  for (const p of props) {
    const tmp = parseFloat(cs.getPropertyValue(p)) || 0;
    outp = tmp > outp ? tmp : outp;
  }
  return outp;
}
