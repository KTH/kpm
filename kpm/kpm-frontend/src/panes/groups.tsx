import React, { useState } from "react";
import { MenuPane, MenuPaneHeader } from "../components/menu";
import { APIGroups } from "kpm-backend-interface";
import { createApiUri, formatTerm, useDataFecther } from "./utils";
import {
  EmptyPlaceholder,
  ErrorMessage,
  LoadingPlaceholder,
} from "../components/common";
import { i18n } from "../i18n/i18n";
import "./groups.scss";
import { IconSettings, IconStar } from "../components/icons";
import { FilterOption, TabFilter } from "../components/filter";

export async function loaderStudies({ request }: any = {}): Promise<APIGroups> {
  const res = await fetch(createApiUri("/api/groups"), {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

type TFilter = "favs" | "all";

export function Groups() {
  const { res, loading, error } = useDataFecther<APIGroups>(loaderStudies);
  const { groups, group_search_url } = res || {};

  const [filter, setFilter] = useState<TFilter>("favs");

  const filteredGroups = groups?.filter((group) => {
    switch (filter) {
      case "favs":
        return group.starred;
      case "all":
        return true;
    }
  });

  const isEmpty = !loading && !error && filteredGroups?.length === 0;

  return (
    <MenuPane className="kpm-groups">
      <MenuPaneHeader title={i18n("My Groups")}>
        <a
          title="Search for interesting groups to follow"
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
        >
          {i18n("Find groups")}
        </a>
        <a href="https://www.kth.se/social/group-create/">
          {i18n("Create new group")}
        </a>
        <IconSettings href="https://www.kth.se/social/home/settings/groups" />
      </MenuPaneHeader>
      <TabFilter>
        <FilterOption<TFilter>
          value="favs"
          filter={filter}
          onSelect={setFilter}
        >
          {i18n("Favourites")}
        </FilterOption>
        <FilterOption<TFilter> value="all" filter={filter} onSelect={setFilter}>
          {i18n("All Subscriptions")}
        </FilterOption>
      </TabFilter>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {isEmpty && (
        <EmptyPlaceholder>
          {filter === "favs"
            ? i18n("You have no groups marked as favourites.")
            : i18n("You don't belong to any groups.")}
        </EmptyPlaceholder>
      )}
      {!isEmpty && (
        <ul>
          {filteredGroups?.map((group) => (
            <li>
              <IconStar className={group.starred ? "star active" : "star"} />
              <a href={group.url}>{group.name}</a>
            </li>
          ))}
        </ul>
      )}
    </MenuPane>
  );
}
