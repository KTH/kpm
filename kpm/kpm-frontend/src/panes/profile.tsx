import * as React from "react";
import { useLoaderData } from "react-router-dom";
import { ErrorMessage, LoadingPlaceholder } from "../components/common";
import { MenuPane } from "../components/menu";
import { createApiUri, useDataFecther } from "./utils";
import { i18n } from "../i18n/i18n";

export async function loaderProfile({ request }: any = {}) {
  const res = await fetch("/kpm/api", {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

export function Profile() {
  const { res, loading, error } = useDataFecther<{ msg: string }>(
    loaderProfile
  );
  const { msg } = res || {};
  // const { msg } = useLoaderData() as { msg: string };
  return (
    <MenuPane>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {msg && (
        <h2>
          {i18n("Profile")} {msg}
        </h2>
      )}
      {msg && (
        <p>
          <a href={createApiUri("/auth/logout")}>{i18n("Logout")}</a>
        </p>
      )}
    </MenuPane>
  );
}
