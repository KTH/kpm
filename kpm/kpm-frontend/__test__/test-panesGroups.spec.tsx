import * as React from "react";
import { describe, expect, test, afterEach } from "@jest/globals";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { Groups } from "../src/panes/groups";
import { mockFetchJson, resetMockedFetch } from "fetch-for-jest";
import { APIGroups } from "kpm-backend-interface";

describe("<Groups />", () => {
  afterEach(() => {
    resetMockedFetch();
  });

  test("Can be rendered", () => {
    const component = renderer.create(<Groups />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Can be rendered with data and none starred", async () => {
    mockFetchJson<APIGroups>({
      groups: [
        {
          name: "Test Group 1",
          url: "//:test/group_1",
          slug: "test_group_1",
          starred: false,
        },
        {
          name: "Test Group 2",
          url: "//:test/group_2",
          slug: "test_group_2",
          starred: false,
        },
      ],
      group_search_url: "https://",
    });

    let component: any;
    await act(() => {
      component = renderer.create(<Groups />);
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Can be rendered with data and one starred", async () => {
    mockFetchJson<APIGroups>({
      groups: [
        {
          name: "Test Group 1 starred",
          url: "//:test/group_1",
          slug: "test_group_1_starred",
          starred: true,
        },
        {
          name: "Test Group 2 not starred",
          url: "//:test/group_2",
          slug: "test_group_2",
          starred: false,
        },
      ],
      group_search_url: "https://",
    });

    let component: any;
    await act(() => {
      component = renderer.create(<Groups />);
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
