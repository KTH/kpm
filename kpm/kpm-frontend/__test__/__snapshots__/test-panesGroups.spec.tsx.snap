// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Groups /> Can be rendered 1`] = `
<dialog
  className="kpm-modal"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-groups"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        My Groups
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <a
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
          title="Search for interesting groups to follow"
        >
          Find groups
        </a>
        <a
          href="https://www.kth.se/social/group-create/"
        >
          Create new group
        </a>
        <a
          className="icon"
          href="https://www.kth.se/social/home/settings/groups"
        >
          <svg
            height="1.2em"
            viewBox="0 0 86 86"
            width="1.2em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Select how you are notified, how the schedule is displayed, or remove subscriptions.
            </title>
            <path
              d="M11.41 72.18a42.98 42.98 0 0 0 20.65 12.41A10.62 10.62 0 0 1 32 83.5a10.5 10.5 0 1 1 20.91 1.35 42.95 42.95 0 0 0 21.3-12.28 10.5 10.5 0 0 1 10.3-18.3 43.05 43.05 0 0 0 0-22.54 10.5 10.5 0 0 1-10.3-18.3A42.94 42.94 0 0 0 52 .94l.01.56a10.5 10.5 0 0 1-21 .2 43 43 0 0 0-19.59 12.12A10.49 10.49 0 0 1 1.38 32.16a43.07 43.07 0 0 0 0 21.68 10.5 10.5 0 0 1 10.04 18.34zM42 57.94a15 15 0 1 0 0-30 15 15 0 0 0 0 30z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </a>
      </aside>
    </header>
    <nav
      className="kpm-tab-filter"
    >
      <ul>
        <li>
          <button
            onClick={[Function]}
          >
            Favourites
          </button>
        </li>
        <li
          className="active"
        >
          <button
            onClick={[Function]}
          >
            All Subscriptions
          </button>
        </li>
      </ul>
    </nav>
    <div
      className="kpm-loading-placeholder"
    >
      <div
        className="kpm-loader"
      />
    </div>
    <ul />
  </div>
</dialog>
`;

exports[`<Groups /> Can be rendered with data and none starred 1`] = `
<dialog
  className="kpm-modal active"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-groups"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        My Groups
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <a
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
          title="Search for interesting groups to follow"
        >
          Find groups
        </a>
        <a
          href="https://www.kth.se/social/group-create/"
        >
          Create new group
        </a>
        <a
          className="icon"
          href="https://www.kth.se/social/home/settings/groups"
        >
          <svg
            height="1.2em"
            viewBox="0 0 86 86"
            width="1.2em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Select how you are notified, how the schedule is displayed, or remove subscriptions.
            </title>
            <path
              d="M11.41 72.18a42.98 42.98 0 0 0 20.65 12.41A10.62 10.62 0 0 1 32 83.5a10.5 10.5 0 1 1 20.91 1.35 42.95 42.95 0 0 0 21.3-12.28 10.5 10.5 0 0 1 10.3-18.3 43.05 43.05 0 0 0 0-22.54 10.5 10.5 0 0 1-10.3-18.3A42.94 42.94 0 0 0 52 .94l.01.56a10.5 10.5 0 0 1-21 .2 43 43 0 0 0-19.59 12.12A10.49 10.49 0 0 1 1.38 32.16a43.07 43.07 0 0 0 0 21.68 10.5 10.5 0 0 1 10.04 18.34zM42 57.94a15 15 0 1 0 0-30 15 15 0 0 0 0 30z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </a>
      </aside>
    </header>
    <nav
      className="kpm-tab-filter"
    >
      <ul>
        <li>
          <button
            onClick={[Function]}
          >
            Favourites
          </button>
        </li>
        <li
          className="active"
        >
          <button
            onClick={[Function]}
          >
            All Subscriptions
          </button>
        </li>
      </ul>
    </nav>
    <ul>
      <li
        className="kpm-groups-item starrable"
      >
        <button
          className="icon star"
          onClick={[Function]}
        >
          <svg
            viewBox="0 0 20 19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10,13.39l-3.76,2.27,.99-4.28-3.32-2.88,4.38-.37,1.71-4.04,1.71,4.04,4.38,.37-3.32,2.88,.99,4.28m6.24-8.42l-7.19-.61L10,0l-2.81,6.63L0,7.24l5.45,4.73-1.63,7.03,6.18-3.73,6.18,3.73-1.64-7.03s5.46-4.73,5.46-4.73Z"
              fill="currentColor"
            />
          </svg>
        </button>
        <a
          href="//:test/group_1"
        >
          Test Group 1
        </a>
      </li>
      <li
        className="kpm-groups-item starrable"
      >
        <button
          className="icon star"
          onClick={[Function]}
        >
          <svg
            viewBox="0 0 20 19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10,13.39l-3.76,2.27,.99-4.28-3.32-2.88,4.38-.37,1.71-4.04,1.71,4.04,4.38,.37-3.32,2.88,.99,4.28m6.24-8.42l-7.19-.61L10,0l-2.81,6.63L0,7.24l5.45,4.73-1.63,7.03,6.18-3.73,6.18,3.73-1.64-7.03s5.46-4.73,5.46-4.73Z"
              fill="currentColor"
            />
          </svg>
        </button>
        <a
          href="//:test/group_2"
        >
          Test Group 2
        </a>
      </li>
    </ul>
  </div>
</dialog>
`;

exports[`<Groups /> Can be rendered with data and one starred 1`] = `
<dialog
  className="kpm-modal active"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-groups"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        My Groups
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <a
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
          title="Search for interesting groups to follow"
        >
          Find groups
        </a>
        <a
          href="https://www.kth.se/social/group-create/"
        >
          Create new group
        </a>
        <a
          className="icon"
          href="https://www.kth.se/social/home/settings/groups"
        >
          <svg
            height="1.2em"
            viewBox="0 0 86 86"
            width="1.2em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Select how you are notified, how the schedule is displayed, or remove subscriptions.
            </title>
            <path
              d="M11.41 72.18a42.98 42.98 0 0 0 20.65 12.41A10.62 10.62 0 0 1 32 83.5a10.5 10.5 0 1 1 20.91 1.35 42.95 42.95 0 0 0 21.3-12.28 10.5 10.5 0 0 1 10.3-18.3 43.05 43.05 0 0 0 0-22.54 10.5 10.5 0 0 1-10.3-18.3A42.94 42.94 0 0 0 52 .94l.01.56a10.5 10.5 0 0 1-21 .2 43 43 0 0 0-19.59 12.12A10.49 10.49 0 0 1 1.38 32.16a43.07 43.07 0 0 0 0 21.68 10.5 10.5 0 0 1 10.04 18.34zM42 57.94a15 15 0 1 0 0-30 15 15 0 0 0 0 30z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </a>
      </aside>
    </header>
    <nav
      className="kpm-tab-filter"
    >
      <ul>
        <li
          className="active"
        >
          <button
            onClick={[Function]}
          >
            Favourites
          </button>
        </li>
        <li>
          <button
            onClick={[Function]}
          >
            All Subscriptions
          </button>
        </li>
      </ul>
    </nav>
    <ul>
      <li
        className="kpm-groups-item starrable active"
      >
        <button
          className="icon star"
          onClick={[Function]}
        >
          <svg
            className="active"
            viewBox="0 0 20 19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10,15.27l6.18,3.73-1.64-7.03,5.46-4.73-7.19-.62L10,0l-2.81,6.62L0,7.24l5.45,4.73-1.63,7.03,6.18-3.73Z"
              fill="currentColor"
            />
          </svg>
        </button>
        <a
          href="//:test/group_1"
        >
          Test Group 1 starred
        </a>
      </li>
    </ul>
  </div>
</dialog>
`;
